src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+            foreach (var protocol in address.Protocols)
+                if (TransportRegistry.Transports.TryGetValue(protocol.Name, out var transport))
+                {
+                    address = transport().Listen(address, OnRemoteConnect, cancel.Token);
+                    _listeners.TryAdd(address, cancel);
+                    didSomething = true;
+                    break;
+                }
+
+            if (!didSomething)
+            {
+                throw new ArgumentException($"Missing a transport protocol name '{address}'.", nameof(address));
+            }
+
+            var result = new MultiAddress($"{address}/ipfs/{LocalPeer.Id}");
+
+            // Get the actual IP address(es).
+            IEnumerable<MultiAddress> addresses = new List<MultiAddress>();
+            var ips = NetworkInterface.GetAllNetworkInterfaces()
+
+                // It appears that the loopback adapter is not UP on Ubuntu 14.04.5 LTS
+               .Where(nic => nic.OperationalStatus == OperationalStatus.Up
+                 || nic.NetworkInterfaceType == NetworkInterfaceType.Loopback)
+               .SelectMany(nic => nic.GetIPProperties().UnicastAddresses);
+            if (result.ToString().StartsWith("/ip4/0.0.0.0/"))
+            {
+                addresses = ips
+                   .Where(ip => ip.Address.AddressFamily == AddressFamily.InterNetwork)
+                   .Select(ip => new MultiAddress(result.ToString().Replace("0.0.0.0", ip.Address.ToString())))
+                   .ToArray();
+            }
+            else if (result.ToString().StartsWith("/ip6/::/"))
+            {
+                addresses = ips
+                   .Where(ip => ip.Address.AddressFamily == AddressFamily.InterNetworkV6)
+                   .Select(ip => { return new MultiAddress(result.ToString().Replace("::", ip.Address.ToString())); })
+                   .ToArray();
+            }
+            else
+            {
+                addresses = new[] {result};
+            }
+            
+            if (!addresses.Any())
+            {
+                var msg = "Cannot determine address(es) for " + result;
+                
+                foreach (var ip in ips)
+                {
+                    msg += " nic-ip: " + ip.Address;
+                }
+                
+                cancel.Cancel();
+                throw new Exception(msg);
+            }
+
+            // Add actual addresses to listeners and local peer addresses.
+            foreach (var a in addresses)
+            {
+                _log.Debug($"Listening on {a}");
+                _listeners.TryAdd(a, cancel);
+            }
+
+            LocalPeer.Addresses = LocalPeer
+               .Addresses
+               .Union(addresses)
+               .ToArray();
+
+            ListenerEstablished?.Invoke(this, LocalPeer);
+            return Task.FromResult(addresses.First());
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not suppress issues.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)